pkg/network/encoding/http.go

```diff
@@ -0,0 +1,130 @@
+// Unless explicitly stated otherwise all files in this repository are licensed
+// under the Apache License Version 2.0.
+// This product includes software developed at Datadog (https://www.datadoghq.com/).
+// Copyright 2016-present Datadog, Inc.
+
+package encoding
+
+import (
+	model "github.com/DataDog/agent-payload/v5/process"
+	"github.com/DataDog/datadog-agent/pkg/network"
+	"github.com/DataDog/datadog-agent/pkg/network/http"
+	"github.com/gogo/protobuf/proto"
+)
+
+var sentinelValue = new(model.HTTPAggregations)
+
+type httpEncoder struct {
+	aggregations map[http.Key]*model.HTTPAggregations
+
+	// pre-allocated objects
+	dataPool []model.HTTPStats_Data
+	ptrPool  []*model.HTTPStats_Data
+	poolIdx  int
+
+	orphanEntries int
+}
+
+func newHTTPEncoder(payload *network.Connections) *httpEncoder {
+	if len(payload.HTTP) == 0 {
+		return nil
+	}
+
+	encoder := &httpEncoder{
+		aggregations: make(map[http.Key]*model.HTTPAggregations, len(payload.Conns)),
+
+		// pre-allocate all data objects at once
+		dataPool: make([]model.HTTPStats_Data, len(payload.HTTP)*http.NumStatusClasses),
+		ptrPool:  make([]*model.HTTPStats_Data, len(payload.HTTP)*http.NumStatusClasses),
+		poolIdx:  0,
+	}
+
+	// pre-populate aggregation map with keys for all existent connections
+	// this allows us to skip encoding orphan HTTP objects that can't be matched to a connection
+	for _, conn := range payload.Conns {
+		encoder.aggregations[httpKeyFromConn(conn)] = sentinelValue
+	}
+
+	encoder.buildAggregations(payload)
+	return encoder
+}
+
+func (e *httpEncoder) GetHTTPAggregations(c network.ConnectionStats) *model.HTTPAggregations {
+	if e == nil {
+		return nil
+	}
+
+	key := httpKeyFromConn(c)
+	if aggregation, ok := e.aggregations[key]; ok && aggregation != sentinelValue {
+		return aggregation
+	}
+
+	return nil
+}
+
+func (e *httpEncoder) buildAggregations(payload *network.Connections) {
+	for key, stats := range payload.HTTP {
+		path := key.Path
+		method := key.Method
+		key.Path = ""
+		key.Method = http.MethodUnknown
+
+		aggregation, ok := e.aggregations[key]
+		if !ok {
+			// if there is no matching connection don't even bother to serialize HTTP data
+			e.orphanEntries++
+			continue
+		}
+
+		if aggregation == sentinelValue {
+			// if this is sentinel value, instantiate a real one
+			aggregation = &model.HTTPAggregations{
+				EndpointAggregations: make([]*model.HTTPStats, 0, 10),
```

Comment:
We can visit this in another PR, but I wonder what kind of gains we could get from tweaking the `10` here. It seems like a lot of speculative pre-allocation.